Readme.md
# xkcd cli tool

<p align="center">
<img alt="Lint and static code analysis on develop branch" src="https://github.com/dotcs/xkcd-cli/actions/workflows/lint-sca.yaml/badge.svg?branch=develop"/>
<a href="https://github.com/dotcs/xkcd-cli/blob/main/LICENSE"><img alt="License: MIT" src="https://black.readthedocs.io/en/stable/_static/license.svg"/></a>
<a href="https://github.com/psf/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg"/></a>
</p>

Get your daily dose of [xkcd] directly from the terminal! 🤩

https://user-images.githubusercontent.com/3976183/163873282-f586f312-2643-4b77-af79-89e344091b2f.mp4

[xkcd] is a webcomic created by [Randall Munroe][munroe]. 
It is a comic of Language, Math, Romance and Sarcasm and a [couple of other categories][explain-xkcd-categories].

If [kitty], [iterm] or any terminal that has support for the [sixel] file format is used as the terminal, the xkcd comic will be rendered directly in the terminal, otherwise the default viewer for PNG images is used.
This tool requires [fzf] to be installed on the machine to filter available comics by their title. 

## Installation

### With pip

Install this package directly from the [Python Package Index (PyPI)][pypi-repo].
The CLI tool requires Python >= 3.8 to be installed.

```console
$ pip install dcs-xkcd-cli
```

This will install a CLI tool named `xkcd` which can be used as described below.

### With pipx

Installation with [pipx] is similar to the pip variant above, but uses `pipx` instead of `pip`.

```console
$ pipx install dcs-xkcd-cli
```

Note that with pipx, this package can be tried out without the need to install it permanently.

```console
$ pipx run dcs-xkcd-cli <args>
```


## Usage

### Search by title

```console
$ xkcd show
```

This functionality requires [fzf] to be installed.

### Show latest xkcd comic

```console
$ xkcd show --latest
```

### Show random xkcd comic

```console
$ xkcd show --random
```

### Show xkcd comic by its ID

```console
$ xkcd show --comic-id 207
```

### Upscaling / width of comics

By default images are upscaled to match the terminal dimensions.
This behavior can be controlled with the `--terminal-scale-up / --no-terminal-scale-up` options.
Images can be also rendered with an explicit width by using the `--width` CLI option.

```console
$ xkcd show --comic-id 207 --no-terminal-scale-up    # disable scaling
$ xkcd show --comic-id 207 --width 1200              # set explicit width
```


### Disable rendering in terminals

```console
$ xkcd show --no-terminal-graphics
```

This command will disable the automatic image protocol detection and directly open the image with the help of `xdg-open` in the default image viewer.

### Disable or update cache

Under the hood this tool uses a cache which is updated once per day transparently.
The cache is used to remember the list of xkcd comics from the [archive].

To disable the cache, use the following command

```console
$ xkcd show --no-cache
```

To update the cache manually, use the following command
```console
$ xkcd update-cache
```

## Development

This repository manages Python dependencies with [poetry].
To install the package and its dependencies run:

```console
$ poetry install
```

The code is formatted with [black] and type checked with [pyright].

Then run the the following commands to lint and test the code:

```console
$ poetry run python -m black --check --diff .   # tests for any lint issues
$ poetry run python -m black .                  # auto-formats the code

$ poetry run python -m pyright                  # runs static code analysis

$ poetry run python -m pytest --cov="xkcd_cli/" --cov-report term --cov-report html   # run tests with code coverage report
```


[fzf]: https://github.com/junegunn/fzf
[kitty]: https://sw.kovidgoyal.net/kitty/
[archive]: https://xkcd.com/archive/
[xkcd]: https://xkcd.com
[munroe]: https://en.wikipedia.org/wiki/Randall_Munroe
[explain-xkcd-categories]: https://www.explainxkcd.com/wiki/index.php/Category:Comics_by_topic
[pypi-repo]: https://pypi.org/project/dcs-xkcd-cli/
[pipx]: https://pypa.github.io/pipx/
[iterm]: https://iterm2.com/
[sixel]: https://en.wikipedia.org/wiki/Sixel
[poetry]: https://python-poetry.org/
[black]: https://black.readthedocs.io/en/stable/
[pyright]: https://github.com/Microsoft/pyright


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2022-06-20

### Added

- [#3](https://github.com/dotcs/xkcd-cli/pull/3): Multiple terminals and image protocols are now auto-detected.
  This includes kitty, iterm and terminals that support the [`sixel` graphic format](https://en.wikipedia.org/wiki/Sixel).
  Thanks to [MatanZ](https://github.com/MatanZ) for this contribution.
- `--width` CLI option to explicitly set the width of the rendered graphic

### Breaking changes

- BREAKING-CHANGE: Removed CLI option `--use-kitty` in favor of `--terminal-graphics`.
- BREAKING-CHANGE: Removed CLI option `--kitty-scale-up` in favor of `--terminal-scale-up`.


## [1.0.0] - 2022-04-17
### Added

- Fetching and caching comics from the xkcd.com archive
- Support for rendering images directly in kitty through its [`icat` command](https://sw.kovidgoyal.net/kitty/kittens/icat/)
- Add support for rendering comics with fuzzy search throught its titles (default), directly (via `--comic-id` option) or at random choice (via `--random`).

xkcd_cli/xkcd.py
from bs4 import BeautifulSoup, Tag
from dataclasses import dataclass, asdict
from datetime import datetime, timedelta
from pathlib import Path
from random import randint
from subprocess import Popen, PIPE
from textwrap import wrap
from typing import List, Tuple, Optional
import json
import os
import requests
import shutil
import subprocess
import sys
import tempfile
import typer
from .iv import IV

BASE_URL = "https://xkcd.com"
ARCHIVE_ENDPOINT = "/archive/"
CACHE_DIR = Path("~", ".cache", "xkcd-cli").expanduser()
CACHE_PATH = Path(CACHE_DIR, "cache.json")
TERM_MAX_WIDTH_CHARS = 80

app = typer.Typer()


@app.callback()
def callback():
    """
    xkcd terminal viewer

    This tool fetches xkcd comics from upstream and allows users to grep a comic
    by selecting it in the terminal.

    It relies on fzf for fuzzy searching the comic titles and a terminal with support
    for graphics to render the images in the terminal.
    """
    pass


def setup():
    os.makedirs(CACHE_DIR, exist_ok=True)


@dataclass
class XkcdComicMeta:
    """
    Meta information of a xkcd comic as described in the archive.
    """

    id: int
    href: str
    title: str


@dataclass
class XkcdComic(XkcdComicMeta):
    """
    Full information of a xkcd comic.
    """

    img_src: str
    subtext: str


@dataclass
class Cache:
    """
    Internal representation of all xkcd comics metadata from the official archive.
    """

    last_updated: datetime
    comics: List[XkcdComicMeta]

    @classmethod
    def read(cls, path: Path):
        content = json.load(open(path))
        last_updated = datetime.fromisoformat(content["last_updated"])
        comics = [XkcdComicMeta(**c) for c in content["comics"]]
        return Cache(last_updated=last_updated, comics=comics)

    def write(self, path: Path):
        with open(path, "w") as f:
            f.write(json.dumps(asdict(self), default=str))


def fetch_xkcd_archive() -> List[XkcdComicMeta]:
    """
    Fetches all xkcd comic meta information from the archive
    page: https://xkcd.com/archive/
    Since a single page contains all historic xkcds, no pagination logic is needed.
    """
    r = requests.get(BASE_URL + ARCHIVE_ENDPOINT)
    r.raise_for_status()
    text = r.text

    soup = BeautifulSoup(text, "html.parser")
    comics: List[XkcdComicMeta] = []

    container = soup.find(id="middleContainer")
    assert isinstance(container, Tag)

    for el in container.find_all("a"):
        href = el.get("href")
        title = el.get_text()
        cid = int(href.replace("/", ""))

        meta = XkcdComicMeta(id=cid, href=href, title=title)
        comics.append(meta)

    return comics


def fetch_xkcd_comic(comic: XkcdComicMeta) -> XkcdComic:
    """
    Fetch an individual xkcd comic from upstream.
    """
    r = requests.get(BASE_URL + comic.href)
    r.raise_for_status()
    html = r.text
    soup = BeautifulSoup(html, "html.parser")
    comic_soup = soup.find(id="comic")
    assert isinstance(comic_soup, Tag)
    img = comic_soup.find("img")
    assert isinstance(img, Tag)
    img_src = img.get("src")
    assert isinstance(img_src, str)
    img_src = "https:" + img_src
    subtext = img.get("title")
    assert isinstance(subtext, str)
    return XkcdComic(
        title=comic.title,
        id=comic.id,
        href=comic.href,
        img_src=img_src,
        subtext=subtext,
    )


def choice_fzf(fzf_cmd: Path, comics: List[XkcdComicMeta]) -> Tuple[bytes, bytes]:
    """
    Uses fzf to select a comic by its title.
    """
    titles = [str(c.id) + ": " + c.title for c in comics]
    content = "\n".join(titles)

    p = Popen([fzf_cmd], stdin=PIPE, stdout=PIPE)
    assert p.stdin
    p.stdin.write(content.encode())
    p.stdin.flush()
    stdout, stderr = p.communicate()
    return stdout, stderr


def _update_cache_if_outdated(
    cache_filename: Path,
    cache_timeout: timedelta = timedelta(hours=24),
) -> Cache:
    """
    Updates the cache if it is older than the last updated timestamp plus the
    cache_timeout.
    """
    cache = Cache.read(cache_filename)
    if datetime.utcnow() > (cache.last_updated + cache_timeout):
        # Transparently create cache for the user if the cache is older than the
        # cache timeout.
        cache = _update_cache(cache_filename)
    return cache


def _update_cache(cache_filename: Path) -> Cache:
    """
    Updates the cache by fetching all comics from the xkcd archive page and
    updating the cache file on disk.
    """
    dt = datetime.utcnow()
    comics = fetch_xkcd_archive()

    cache = Cache(last_updated=dt, comics=comics)
    cache.write(cache_filename)
    return cache


@app.command()
def update_cache(
    cache_filename: Path = typer.Option(
        CACHE_PATH,
        writable=True,
        help="Path to the cache file.",
    )
):
    """
    Updates the cache file by fetching the latest comics from the xkcd archive website.
    """
    _update_cache(cache_filename)
    typer.echo("Cache updated 👍")


@app.command()
def show(
    terminal_graphics: bool = typer.Option(
        True,
        help=(
            "Defines if the output will be displayed in terminal by "
            "rendering the image in the terminal window if supported."
        ),
    ),
    fzf_cmd: Path = typer.Option(
        shutil.which("fzf") or os.getenv("FZF_CMD"),
        help="Path to the fzf tool",
    ),
    width: int = typer.Option(
        -1,
        help="Defines the target width of the comic when rendered in the terminal.",
    ),
    terminal_scale_up: bool = typer.Option(
        True,
        help=(
            "Scales the image up to max possible width if terminal graphics is being "
            "used. Does not have any effect if 'width' is set to an explicit value."
        ),
    ),
    latest: bool = typer.Option(
        False,
        help=(
            "Fetches and renders the latest xkcd without going through a selection first."
        ),
    ),
    random: bool = typer.Option(False, help="Fetches and renders a random xkcd comic."),
    comic_id: Optional[int] = typer.Option(
        None,
        help="Renders a comic with a certain ID.",
    ),
    cache: bool = typer.Option(
        True,
        help=(
            "Defines if a cache should be used for listing all available xkcd "
            "comics. Otherwise calls the xkcd archive endpoint to gather the list "
            "of comics (slower)."
        ),
    ),
    cache_filename: Path = typer.Option(
        CACHE_PATH,
        exists=False,
        readable=True,
        help="Path to the cache file.",
    ),
):
    """
    Show an individual xkcd comic.
    """
    iv: Optional[IV] = None
    if terminal_graphics:
        iv = IV("auto")
        if not iv.protocol:
            terminal_graphics = False
    if cache:
        if not cache_filename.exists():
            # Transparently create cache for the user if cache does not yet exist.
            _update_cache(cache_filename)

        content = _update_cache_if_outdated(cache_filename)
        comics = content.comics
    else:
        # Bypass cache and read fetched results directly
        comics = fetch_xkcd_archive()

    if random:
        meta = comics[randint(0, len(comics))]
    elif latest:
        meta = comics[0]
    elif comic_id is not None:
        try:
            meta = next(c for c in comics if c.id == comic_id)
        except StopIteration:
            typer.echo(
                f"""\
Comic with ID {comic_id} is unknown. Sometimes this happens because the cache is \
outdated. Please use the 'update-cache' command to fetch most recent comics from \
xkcd upstream.""",
                err=True,
            )
            sys.exit(1)
    else:
        stdout, _ = choice_fzf(fzf_cmd, comics)
        choice = stdout.decode("UTF-8").strip()
        try:
            choice_title = choice.split(":")[1].strip()
        except IndexError:
            # Happens if user uses ctrl+c to exit selection
            typer.echo("Unknown index. Abort.", err=True)
            sys.exit(1)

        meta = next(c for c in comics if c.title == choice_title)
    comic = fetch_xkcd_comic(meta)

    wrapped_title = "\n".join(wrap(comic.title, width=TERM_MAX_WIDTH_CHARS))
    typer.echo(typer.style(wrapped_title, bold=True) + f" ({comic.id})")

    # Kitty and the alternative method based on xdg-open would support rendering
    # images directly from an HTTP endpoint but we find it cleaner to download
    # the file first and serve it from local disk.
    with tempfile.TemporaryDirectory() as tempdir:
        r = requests.get(comic.img_src)
        r.raise_for_status()
        tmp_img_path = Path(tempdir, str(comic.id) + ".png", stream=True)
        with open(tmp_img_path, "wb") as f:
            for chunk in r:
                f.write(chunk)
        if terminal_graphics:
            assert iv is not None  # safe since iv has been initialized above
            fitscreen = (
                width <= 0
            )  # disable fit to screen if explicit width has been set by user
            iv.show_image(
                str(tmp_img_path),
                w=width,
                newline=True,
                fitwidth=fitscreen,
                fitheight=fitscreen,
                upscale=terminal_scale_up,
            )
        else:
            cmd = [
                "xdg-open",
                tmp_img_path,
            ]
            subprocess.run(cmd, stdout=None, stderr=None)
    typer.echo("\n".join(wrap(comic.subtext, width=TERM_MAX_WIDTH_CHARS)))


def main():
    setup()
    app()


if __name__ == "__main__":
    main()


xkcd_cli/iv.py
#!/usr/bin/python3
import shutil
import sys
import termios
import atexit
import select
from base64 import standard_b64encode
from typing import BinaryIO, Literal, Optional, Set, TextIO, Tuple, Union
import subprocess

Protocol = Union[
    Literal["iterm"], Literal["kitty"], Literal["kitty+"], Literal["sixel"]
]


class IV:
    protocols: Set[Protocol] = set(("iterm", "kitty", "kitty+", "sixel"))
    protocol: Optional[Protocol]
    kitty: Optional[bool] = None
    ex_kitty: Optional[bool] = None
    sixel: Optional[bool] = None
    iterm: Optional[bool] = None

    def __init__(self, protocol: Optional[Union[Protocol, Literal["auto"]]] = None):
        self.libsixel = None
        self.stdin_fd = sys.stdin.fileno()
        self.saved_term = termios.tcgetattr(self.stdin_fd)
        atexit.register(self.set_normal_term)
        if protocol == "auto":
            self.auto_protocol()
        elif protocol in self.protocols:
            self.protocol = protocol
        else:
            self.protocol = None

    @staticmethod
    def scale_fit(
        w: int,
        h: int,
        ow: int,
        oh: int,
        aspect: bool = True,
        up: bool = False,
        debug: bool = False,
    ) -> Tuple[int, int]:
        """Utility function which calculates new dimension of image"""
        if debug:
            print(w, h, ow, oh, aspect)
        a = ow / oh
        if not up and w > ow:
            w = ow
        if not up and h > oh:
            h = oh
        if w > 0 and h > 0:
            if aspect:
                if w / h > a:
                    return int(h * a), h
                else:
                    return w, int(w / a)
            else:
                return w, h
        elif w > 0:
            if aspect:
                return w, int(w / a)
            else:
                return w, oh
        elif h > 0:
            if aspect:
                return int(h * a), h
            else:
                return ow, h
        else:
            return ow, oh

    # Send and escape sequence and read reply.
    def set_normal_term(self) -> None:  # pragma: no cover
        """
        Reset the terminal to normal mode.
        """
        termios.tcsetattr(self.stdin_fd, termios.TCSAFLUSH, self.saved_term)

    def set_raw_like_term(self) -> None:  # pragma: no cover
        """
        Sets the terminal in raw-like mode (noncanonical mode + nonecho mode).

        See also: https://docs.python.org/3.8/library/termios.html#termios.tcgetattr
        """
        new_term = termios.tcgetattr(self.stdin_fd)
        new_term[3] = new_term[3] & ~termios.ICANON & ~termios.ECHO  # lflags
        termios.tcsetattr(self.stdin_fd, termios.TCSAFLUSH, new_term)

    def terminal_request(
        self,
        cmd: str,
        end: str,
        out: BinaryIO = sys.stdout.buffer,
        in_: TextIO = sys.stdin,
    ) -> str:
        """
        In raw-like term, run the `cmd` and read from stdin until the matching
        `end` character has been found. After executing the command, the command
        terminal is switched back to normal mode.
        """
        try:
            self.set_raw_like_term()
            ret = ""
            out.write(cmd.encode("ascii"))
            out.flush()
            timeout = 0.2  # in seconds
            dr, _, _ = select.select(
                [in_], [], [], timeout
            )  # wait for response on stdin
            if dr != []:
                while True:
                    c = in_.read(1)
                    ret += c
                    if c in end:
                        break
        finally:
            # ensure to switch back to normal terminal mode
            self.set_normal_term()
        return ret

    # Methods to send show image in various protocols.
    def kitty_remove_placement(self, p: int = -1) -> None:
        """
        Delete an image.
        See kitty documentation for details: https://sw.kovidgoyal.net/kitty/graphics-protocol/#deleting-images
        """
        sys.stdout.buffer.write(f"\033_Ga=d,i=-1,d=z,z={p},q=2\033\\".encode("ascii"))
        sys.stdout.flush()

    @staticmethod
    def image_data_and_metadata(data: bytes) -> Tuple[bytes, Optional[str], int, int]:
        from PIL import Image
        import io

        image = Image.open(io.BytesIO(data))
        if image.format == "PNG":
            data = standard_b64encode(data)
        else:
            data = standard_b64encode(image.tobytes())

        return data, image.format, image.height, image.width

    def kitty_show_file(
        self,
        data: bytes,
        debug: bool = False,
        extended: Optional[bool] = None,
        out: BinaryIO = sys.stdout.buffer,
        **params: int,
    ) -> None:
        if extended or (extended is None and self.ex_kitty):
            # if rendering JPG data is supported, no additional action is required
            data = standard_b64encode(data)
        else:
            # if rendering JPG data is not supported, either use PNG directly or
            # use raw image instead
            data, img_format, img_height, img_width = IV.image_data_and_metadata(data)
            if img_format != "PNG":
                params["f"] = 24  # 24-bit RGB data
                params["s"] = img_width
                params["v"] = img_height

        if debug:
            print(len(data), "bytes", file=sys.stderr)

        # See explanation of kitty control data keys here:
        # https://sw.kovidgoyal.net/kitty/graphics-protocol/?highlight=image#control-data-reference
        cmd = dict()
        cmd["a"] = "T"  # simultaneously transmit and display an image
        cmd["f"] = 100  # f=100 indicates PNG data
        cmd["q"] = 2  # Suppress responses from the terminal to this graphics command
        for k in ("C", "p", "i", "z", "a", "f", "q", "s", "v", "c", "r"):
            if k in params:
                cmd[k] = params[k]

        # Write image to kitty terminal in chunks, `m=1` means there is more
        # chunked data available.
        first = True
        s = 0
        chunk_size = 4096
        while s < len(data):
            w = b"\033_G"
            if first:
                for l, v in cmd.items():
                    w += f"{l}={v},".encode("ascii")
                    first = False
            s += chunk_size
            w += f"m={int(s<len(data))};".encode("ascii")
            w += data[s - chunk_size : s]
            w += b"\033\\"
            out.write(w)
            out.flush()

    def iterm_show_file(
        self,
        data: bytes,
        debug: bool = False,
        out: BinaryIO = sys.stdout.buffer,
        **params: int,
    ):
        data = standard_b64encode(data)
        if debug:
            print(len(data), "bytes", file=sys.stderr)
        extras = ""
        if "C" in params and params["C"] == 1:
            extras += ";doNotMoveCursor=1"
        if "width" in params:
            extras += f";width={params['width']}"
        out.write(
            b"\033]1337;File=inline=1" + bytes(extras, "ascii") + b":" + data + b"\007"
        )
        out.flush()

    def _setup_libsixel_or_fallback(self):  # pragma: no cover
        try:
            from libsixel import encoder  # type: ignore

            self.encoder = encoder
            self.libsixel = True
        except ModuleNotFoundError:
            self.libsixel = False

    def sixel_show_file(
        self,
        filename: str,
        w: int = -1,
        h: int = -1,
        out: BinaryIO = sys.stdout.buffer,
    ) -> None:
        if self.libsixel is None:
            self._setup_libsixel_or_fallback()
        if self.libsixel:
            enc = self.encoder.Encoder()
            enc.setopt(self.encoder.SIXEL_OPTFLAG_WIDTH, str(w) if w > 0 else "auto")
            enc.setopt(self.encoder.SIXEL_OPTFLAG_HEIGHT, str(h) if h > 0 else "auto")
            enc.setopt(self.encoder.SIXEL_OPTFLAG_COLORS, "256")
            enc.encode(filename)
        elif shutil.which("convert") is not None:
            # Use imagemagick convert command as a fallback to convert to sixel format.
            # See also: https://konfou.xyz/posts/sixel-for-terminal-graphics/
            command = ["convert", filename]
            if w > 0 or h > 0:
                command += ["-geometry", f"{w if w > 0 else ''}x{h if h > 0 else ''}"]
            command += ["sixel:-"]
            res = subprocess.run(command, stdout=subprocess.PIPE)
            out.write(res.stdout)
            out.flush()
        else:
            out.write("Could not find an terminal image renderer.".encode("ascii"))
            out.flush()

    def show_image(
        self,
        image: Union[str, bytes],
        w: int = -1,
        h: int = -1,
        newline: bool = False,
        fitwidth: bool = False,
        fitheight: bool = False,
        upscale: bool = False,
        **params: int,
    ):
        """
        Display an image in the current terminal instance.

        Uses the following logic:
        1. If `fitwidth` then scale the image (keeping aspect ratio) so it is as
           wide as the terminal.
        2. If `fitheight` then scale the image (keeping aspect ratio) so it is as
            high as the terminal.
        3. If both are given uses the smaller scale of the above.
        4. Otherwise: If `w` is positive, scale the image (keeping aspect ratio)
           to the given width.
        5. If `h` is positive, scale the image (keeping aspect ratio) to the given
           height.
        6. If both are given uses the smaller scale of the above
        7. In all cases if the scaling factor is larger than 1, it is applied
           only if `upscale` is true.
        """
        if self.protocol not in self.protocols:
            return
        params = {}
        if fitwidth:
            _, w = self.terminal_pixel_size()
        if fitheight:
            h, _ = self.terminal_pixel_size()
        if isinstance(image, str):
            self._show_image_str(image, w, h, upscale=upscale, **params)
        elif isinstance(image, bytes):
            self._show_image_bytes(image, w, h, **params)
        if newline:
            print()

    def _show_image_str(
        self,
        image: str,
        w: int = -1,
        h: int = -1,
        upscale: bool = False,
        **params: int,
    ):
        if self.protocol == "sixel":
            self.sixel_show_file(image, w, h)
        else:
            if w > 0 or h > 0:
                from PIL import Image
                import io

                im = Image.open(image)
                x, y = im.size
                nw, nh = IV.scale_fit(w, h, x, y, up=upscale)
                data = io.BytesIO()
                im.resize((nw, nh)).save(data, format=im.format)
                data.seek(0)
                data = data.read()
            else:
                data = open(image, "rb").read()
            if self.protocol == "iterm":
                self.iterm_show_file(data, **params)
            elif self.protocol is not None and self.protocol.startswith("kitty"):
                self.kitty_show_file(data, **params)

    def _show_image_bytes(
        self,
        image: bytes,
        w: int = -1,
        h: int = -1,
        **params: int,
    ):
        if self.protocol == "sixel":
            import tempfile

            file = tempfile.TemporaryFile()
            file.write(image)
            self.sixel_show_file(file.name, w, h)
        else:
            if self.protocol == "iterm":
                self.iterm_show_file(image, **params)
            elif self.protocol is not None and self.protocol.startswith("kitty"):
                self.kitty_show_file(image, **params)

    # Get various sizes of screen
    def terminal_pixel_size(self) -> Tuple[int, int]:
        """
        Use ANSI escape code to determine the terminal screen dimensions in
        pixels of the area that may be covered by graphics.
        See also: https://notes.burke.libbey.me/ansi-escape-codes/
        See also: https://sw.kovidgoyal.net/kitty/graphics-protocol/#getting-the-window-size
        """
        size_ret = self.terminal_request("\x1b[14t", "t").split(";")
        if len(size_ret) < 3 or not size_ret[0][-1] == "4":
            terminal_width, terminal_height = -1, -1
        else:
            terminal_height, terminal_width = int(size_ret[1]), int(size_ret[2][:-1])
        return terminal_height, terminal_width

    def cell_size(self) -> Tuple[int, int]:
        """
        Use ANSI escape code to determine the cell dimensions.
        See also: https://notes.burke.libbey.me/ansi-escape-codes/
        """
        size_ret = self.terminal_request("\x1b[16t", "t").split(";")
        if len(size_ret) < 3 or not size_ret[0][-1] == "6":
            cell_width, cell_height = -1, -1
        else:
            cell_height, cell_width = int(size_ret[1]), int(size_ret[2][:-1])
        return cell_height, cell_width

    def iterm_cell_size(self) -> Tuple[int, int]:
        """
        Use proprietary ANSI escape code to termine cell dimensions in iTerm.
        See also: https://iterm2.com/documentation-escape-codes.html
        """
        size_ret = self.terminal_request(
            "\x1b]1337;ReportCellSize\x07", "\x07\\"
        ).split(";")
        if len(size_ret) < 3 or not size_ret[1].startswith("ReportCellSize="):
            cell_width, cell_height = -1, -1
        else:
            cell_height, cell_width = float(size_ret[1].split("=")[1]), float(
                size_ret[2].strip("\07\\\x1b")
            )
            if len(size_ret) == 4:
                factor = float(size_ret[3].strip("\07\\\x1b"))
            else:
                factor = 1
            cell_height, cell_width = round(cell_height * factor), round(
                cell_width * factor
            )

        return cell_height, cell_width

    def terminal_cell_size(self) -> Tuple[int, int]:
        """
        Use ANSI escape code to determine the terminal dimensions as number of
        lines and columns.
        See also: https://notes.burke.libbey.me/ansi-escape-codes/
        """
        size_ret = self.terminal_request("\x1b[18t", "t").split(";")
        if len(size_ret) < 3 or not size_ret[0][-1] == "8":
            terminal_lines, terminal_cols = -1, -1
        else:
            terminal_lines, terminal_cols = int(size_ret[1]), int(size_ret[2][:-1])
        return terminal_lines, terminal_cols

    # Identify supported protocols
    def have_iterm(self) -> bool:
        """
        Identify if the current terminal is an `iterm` terminal instance by
        running a proprietary command that only works on those terminals.

        This method sets the `iterm` variable on the current instance.
        """
        if self.iterm is None:
            self.iterm = -1 not in self.iterm_cell_size()
        return self.iterm

    def have_kitty(self) -> bool:
        """
        Identify if the current terminal is a `kitty` terminal instance by
        sending a sample 24-bit RGB image with dimensions 1x1 (params s, v) in
        query action mode (a=q).

        This method sets the `kitty` variable on the current instance.

        See also: https://sw.kovidgoyal.net/kitty/graphics-protocol/?highlight=image#querying-support-and-available-transmission-mediums
        """
        if self.kitty is None:
            self.kitty = "OK" in self.terminal_request(
                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=24;AAAA\x1b\\", "\\"
            )
        return self.kitty

    def have_extended_kitty(self) -> bool:
        """
        Determine if extended support for JPG images is available.
        Note that JPG support is implemented by a limited number of terminals,
        e.g. konsole or wezterm, but not kitty.

        This method sets the `ex_kitty` variable, and the `kitty` variable if
        unset, on the currrent instance.
        """
        if self.ex_kitty is None:
            # A small jpeg from https://stackoverflow.com/questions/2253404
            self.ex_kitty = "OK" in self.terminal_request(
                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=100;"
                + "/9j/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wgALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAAA//aAAgBAQAAAAE//9k="
                + "\x1b\\",
                "\\",
            )
            if self.ex_kitty and self.kitty is None:
                self.kitty = True
        return self.ex_kitty

    def have_sixel(self) -> bool:
        """
        Determine if sixel support is available in the current terminal instance
        through ANSI escape code.

        This method sets the `sixel` variable on the current instance.
        """
        if self.sixel is None:
            # reply starts with '\x1b[?' and ends with 'c'
            self.sixel = "4" in self.terminal_request("\x1b[c", "c")[3:-1].split(";")
        return self.sixel

    def auto_protocol(self) -> Optional[Protocol]:
        """
        Automatically detect the terminal image protocol by testing various
        methods on-the-fly.

        This method sets the `protocol` variable on the current instance.
        """
        protocol: Optional[Protocol] = None
        if self.have_extended_kitty():
            protocol = "kitty+"
        elif self.have_iterm():
            protocol = "iterm"
        elif self.have_kitty():
            protocol = "kitty"
        elif self.have_sixel():
            protocol = "sixel"
        self.protocol = protocol
        return protocol
